implants/lib/eldritch/src/pivot/ssh_exec_impl.rs

```diff
@@ -1,5 +1,205 @@
+    let mut ssh = tokio::time::timeout(
+        std::time::Duration::from_secs(timeout.unwrap_or(3).into()),
+        Session::connect(username, password, key, key_password, format!("{}:{}", target, port)),
+
+    ).await??;
+    let r = ssh.call(&command).await?;
+    ssh.close().await?;
+
+    Ok(
+        SSHExecOutput { 
+            stdout: r.output(),
+            status: r.code.unwrap_or(0) as i32,
+        }
+    )
+}
+
+pub fn ssh_exec(starlark_heap: &Heap, target: String, port: i32, command: String, username: String, password: Option<String>, key: Option<String>, key_password: Option<String>, timeout: Option<u32>) -> Result<Dict> {
+    
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let key_password_ref = key_password.as_deref();
+    let local_port: u16 = port.try_into()?;
+
+    let cmd_res = match runtime.block_on(
+        handle_ssh_exec(target, local_port, command, username, password, key, key_password_ref, timeout)
+    ) {
+        Ok(local_res) => local_res,
+        Err(local_err) => return Err(anyhow::anyhow!("Failed to run handle_ssh_exec: {}", local_err.to_string())),
+    };
+
+    let res = SmallMap::new();
+    let mut dict_res = Dict::new(res);
+    let stdout_value = starlark_heap.alloc_str(&cmd_res.stdout);
+    dict_res.insert_hashed(const_frozen_string!("stdout").to_value().get_hashed().unwrap(), stdout_value.to_value());
+
+    let status_value = Value::new_int(cmd_res.status);
+    dict_res.insert_hashed(const_frozen_string!("status").to_value().get_hashed().unwrap(), status_value);
+
+    Ok(dict_res)
+}
+
+#[cfg(test)]
+mod tests {
+    use async_trait::async_trait;
+    use tokio::net::TcpListener;
+    use tokio::task;
+    use std::process::Command;
+    use std::sync::{Mutex, Arc};
+    use russh::*;
+    use russh::server::{Auth, Session, Msg};
+    use russh_keys::*;
+    use std::collections::HashMap;
+    use super::*;
+
+    // SSH Server utils
+    #[derive(Clone)]
+    #[allow(dead_code)]
+    struct Server {
+        client_pubkey: Arc<russh_keys::key::PublicKey>,
+        clients: Arc<Mutex<HashMap<(usize, ChannelId), Channel<Msg>>>>,
+        id: usize,
+    }
+
+    impl server::Server for Server {
+        type Handler = Self;
+        fn new_client(&mut self, _: Option<std::net::SocketAddr>) -> Self {
+            let s = self.clone();
+            self.id += 1;
+            s
+        }
+    }
+
+    #[async_trait]
+    impl server::Handler for Server {
+        type Error = anyhow::Error;
+
+        async fn channel_open_session(self, channel: Channel<Msg>, session: Session) -> Result<(Self, bool, Session), Self::Error> {
+            {
+                let mut clients = self.clients.lock().unwrap();
+                clients.insert((self.id, channel.id()), channel);
+            }
+            Ok((self, true, session))
+        }
+
+        #[allow(unused_variables)]
```

Comment:
sus

Reply:
Yeah same deal ^ library's boiler plate is kinda scuffed.
I don't think any code is actually unused but that's how it lints out.